packages/integration-sdk-runtime/src/synchronization/index.ts

```diff
@@ -426,6 +440,87 @@ export async function uploadData<T extends UploadDataLookup, K extends keyof T>(
   );
 }
 
+/**
+ * Removes data from the rawData of the largest entity until the overall size
+ * of the data object is less than maxSize (defaulted to UPLOAD_SIZE_MAX).
+ *
+ * @param data
+ */
+export function shrinkRawData<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  logger,
+  maxSize = UPLOAD_SIZE_MAX,
+) {
+  const startTimeInMilliseconds = Date.now();
+  let totalSize = Buffer.byteLength(JSON.stringify(data));
+  const initialSize = totalSize;
+  let itemsRemoved = 0;
+  const sizeOfTruncated = Buffer.byteLength("'TRUNCATED'");
+
+  while (totalSize > maxSize) {
+    let largestEntityKey = '';
+    let largestEntitySize = 0;
+    let largestRawDataEntryKey = '';
+    let largestRawDataEntrySize = 0;
+    let largestItemKey = '';
+    let largestItemSize = 0;
+
+    // Find largest Entity
+    for (const entity in data) {
+      const length = JSON.stringify(data[entity]).length;
+      if (length > largestEntitySize) {
+        largestEntitySize = length;
+        largestEntityKey = entity;
+      }
+    }
+
+    // Find largest _rawData entry (typically 0, but check to be certain)
+    for (const rawEntry in data[largestEntityKey]['_rawData']) {
+      const length = JSON.stringify(
+        data[largestEntityKey]['_rawData'][rawEntry],
+      ).length;
+      if (length > largestRawDataEntrySize) {
+        largestRawDataEntrySize = length;
+        largestRawDataEntryKey = rawEntry;
+      }
+    }
+
+    // Find largest item within rawData
+    for (const item in data[largestEntityKey]['_rawData'][
+      largestRawDataEntryKey
+    ]['rawData']) {
+      const length = data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
+        'rawData'
+      ][item]
+        ? Buffer.byteLength(
+            JSON.stringify(
+              data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
+                'rawData'
+              ][item],
+            ),
+          )
+        : 0;
```

Comment:
There's a lot going on here too and I'm having a bit of difficultly parsing all of it in my head 😅 

How about we split this into two functions:

```ts
const entityRawDataEntryByteLength = getByteLengthForEntityRawDataEntry(...)
```

```ts
const largestEntityItemSize = getLargestEntityItemSize(...)
```